[SR-14813] [Fuzzers] Fix the mangler and reflection fuzzer build.

al45tair opened the pull request:
The mangler fuzzer and reflection fuzzer build was broken, both by a problem in the CMake scripts and also because of changes that have happened in other parts of the code.

Resolves SR-14813.
rdar://79623532

al45tair committed:
[SR-14813] [Fuzzers] Fix the mangler and reflection fuzzer build.

The mangler fuzzer and reflection fuzzer build was broken, both by a problem in
the CMake scripts and also because of changes that have happened in other parts
of the code.

al45tair commented:
@swift-ci Please smoke test

gottesmm commented on tools/swift-reflection-fuzzer/swift-reflection-fuzzer.cpp:
```diff
@@ -73,6 +74,15 @@ class ObjectMemoryReader : public MemoryReader {
+    case DLQ_GetPtrAuthMask: {
+      auto result = static_cast<uintptr_t *>(outBuffer);
+#if __has_feature(ptrauth_calls)
+      *result = (uintptr_t)ptrauth_strip((void*)0x0007ffffffffffff, 0);
```
Do we have a true mask that we can rely on instead of hard coding?

gottesmm replied:
Can you add a comment here explaining what you are doing in more detail. That is can you explain here what you are actually testing.

gottesmm replied:
Also you should run this through git-clang-format. The reason why I know that you haven't is that `(void*)` should be `(void *)`. Also can you use static_cast like the rest of the code rather than using c casts?

gottesmm replied:
Also I think that __has_feature is a clang extension.

al45tair replied:
All of the above is true, however I copied this piece of code directly from `include/swift/Remote/InProcessMemoryReader.h` (apparently the original was written by @kubamracek).

Not sure where that leaves us exactly. Should I fix that file at the same time?

al45tair replied:
As far as `__has_feature()` goes, it looks like we use it all over the shop, and various files have their own variant of

    #ifndef __has_feature
    #define __has_feature(x) 0
    #endif

which arguably belongs in `swift/Basic/Compiler.h` with the other similar defines.

gottesmm commented on tools/swift-reflection-fuzzer/swift-reflection-fuzzer.cpp:
```diff
@@ -131,6 +141,6 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
   NativeReflectionContext context(std::move(reader));
   context.addImage(RemoteAddress(Data));
-  context.getBuilder().dumpAllSections(std::cout);
+  context.getBuilder().dumpAllSections(stdout);
```
Did the ABI change here and we just never built it? Why are you making this change here?

al45tair replied:
Apparently it did. `dumpAllSections()` currently takes a `FILE *` and presumably at some point in the past it must have taken an `ostream&` or similar; either that, or the reflection fuzzer never worked, which I find hard to believe.

al45tair replied:
It was changed in dd7d173c97ea "[stdlib] [reflection] Excise upstream from reflection".

gottesmm commented on cmake/modules/AddSwift.cmake:
```diff
@@ -885,7 +885,7 @@ function(add_swift_fuzzer_host_tool executable)
   # Then make sure that we pass the -fsanitize=fuzzer flag both on the cflags
   # and cxx flags line.
-  target_compile_options(${executable} PRIVATE $<$<COMPILE_LANGUAGE:C,CXX,OBJC,OBJCXX>:"-fsanitize=fuzzer">)
+  target_compile_options(${executable} PRIVATE $<$<COMPILE_LANGUAGE:C,CXX,OBJC,OBJCXX>:-fsanitize=fuzzer>)
```
Do you know if swiftc also supports -sanitize=fuzzer? If so, we should add that as well.

gottesmm commented on tools/swift-reflection-fuzzer/swift-reflection-fuzzer.cpp:
```diff
@@ -28,6 +28,7 @@
 #include "llvm/Object/ELFObjectFile.h"
 #include "llvm/Object/MachOUniversal.h"
 #include "llvm/Support/CommandLine.h"
+#include <stdio.h>
```
This is a cpp file. Can you please use `cstdio` and also change stddef/stdint to also use the proper c++ include.

al45tair replied:
Funnily enough, I started using `cstdio` and then assumed that that wasn't the right thing because of the other headers. So yes, let's do that instead.

al45tair committed:
[SR-14813] Add Swift compile options for fuzzing.

We only had the C/C++/ObjC/ObjC++ option.  Add the Swift one too.

al45tair committed:
[SR-14813] Use the C++ standard library includes.

Replace use of stdio.h, stddef.h and stdint.h with the C++ versions of
those headers.

al45tair committed:
[SR-14813] Tidy up the DLQ_GetPtrAuthMask case.

We should be using C++ casts here.

al45tair committed:
[SR-14813] Reformat code slightly.

Used git-clang-format to fix up the changed code.

al45tair commented:
@swift-ci Please smoke test
